### ISSUE TITLE
[ANTLR] Some nullable types are inheritable.

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/sdk#51706]
There are several places where `typeNotVoidNotFunction` makes it possible for nullable types to be extended/mixed-in/implemented or be part of a `mixinApplication`.

It looks like extending nullable types is not intended to be supported and is a compile-time error because the following does not compile:

```dart
//   A class can't extend a nullable type.
//                vvvvvvvvv
class Foo extends Function? {}
```

I think there is a simple adjustment that would allow the grammar to specify that some places don't expect nullable types by making them a syntax error i.e. by replacing the [`typeNotVoidNotFunction`](https://github.com/dart-lang/sdk/blob/6c7f3ab27d68268dcbe34cbb805b324094fe508d/tools/spec_parser/Dart.g#L1409-L1413) where nullable types are not expected with one that does not have trailing questionmarks.

I found the following places that look like they are not intended to expect nullable types:

https://github.com/dart-lang/sdk/blob/6c7f3ab27d68268dcbe34cbb805b324094fe508d/tools/spec_parser/Dart.g#L557-L559

https://github.com/dart-lang/sdk/blob/6c7f3ab27d68268dcbe34cbb805b324094fe508d/tools/spec_parser/Dart.g#L411-L415

and via `typeNotVoidNotFunctionList`

https://github.com/dart-lang/sdk/blob/6c7f3ab27d68268dcbe34cbb805b324094fe508d/tools/spec_parser/Dart.g#L398-L400

https://github.com/dart-lang/sdk/blob/6c7f3ab27d68268dcbe34cbb805b324094fe508d/tools/spec_parser/Dart.g#L394-L396

https://github.com/dart-lang/sdk/blob/6c7f3ab27d68268dcbe34cbb805b324094fe508d/tools/spec_parser/Dart.g#L389-L392

The LaTeX-based spec seems to mirror the behavior of the ANTLR-based spec (unless it specifies that this is not allowed in natural language). See: [typeNotVoidNotFunction](https://github.com/dart-lang/language/blob/2a3a253ca81d3cf0b78ad88e34aaaf6e185b5345/specification/dartLangSpec.tex#L20189-L20190) and e.g. [superclass](https://github.com/dart-lang/language/blob/2a3a253ca81d3cf0b78ad88e34aaaf6e185b5345/specification/dartLangSpec.tex#L4502-L4504)

**Comment by eernstg** [comment]
There is already a compile-time error for this [here](https://github.com/dart-lang/language/blob/master/accepted/2.12/nnbd/feature-specification.md#:~:text=It%20is%20an-,error,-for%20a%20class%20to%20extend%2C%20implement%2C%20or%20mixin%20a). Note that the type `T?` may be specified explicitly, and it may also be defined via one or more type aliases:

```dart
class A {}
typedef MaybeA = A?;
class B extends MaybeA {} // Error.
```

Hence, we wouldn't be able to catch more than a subset of these situations if we were to express the constraint in the grammar, and then we would still have to have the compile-time error that we have today (that is, a compile-time error which is not a syntax error). So, in this case it isn't useful to change the grammar.

**Comment by modulovalue** [comment]
I agree with you. I think we just have different points of view (and I think both are valid) which are:

- Do those places accept actual types or
- Do those places accept constructs that refer to some declarations (that just so happen to look like types).

You seem to favor the first point of view, I'm suggesting that we should take the second point of view. 

> Another thing worth considering is that the exceptions for void and for function types 

and for record types. Only type references and a function literal are left.

The function literal there is deprecated: [deprecated_subtype_of_function](https://dart.dev/tools/diagnostic-messages#deprecated_subtype_of_function) and once `'Function'`s have been removed from being supported in those places, we will be able to simplify typeNotVoidNotFunction even further to e.g.:

```dart
typeReference : typeName typeArguments?
```

That's why I think the second point of view will make things slightly simpler here and should be preferred.

(_Of course, implementations of the spec could choose to accept full types and not just typeReferences for e.g. error recovery purposes_)

> and there is no good reason for having those exceptions in the grammar, given that they must again be enforced using a non-syntax compile-time error as well.

With the proposed changes, the enforcement will need to happen on what typeReference **refers to**, and not the construct that **does the actual referencing** because it would no longer be a type that needs to be validated, but a reference. And guaranteed to be correct by the grammar (i.e. definitely no `?`).

**Comment by modulovalue** [comment]
Closed via #51742
